File: examples/javascript/tests/circle.test.js
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { Circle } from "../src/circle.js";
import { Color }  from "../src/shape.js";
import { Vector3 } from "../src/vector3.js";

describe("Circle", () => {
  describe("defaults", () => {
    it("uses BLUE as default color", () => {
      const c = new Circle();
      assert.equal(c.color, Color.BLUE);
    });

    it("defaults center to (0,0,0) and radius to 1", () => {
      const c = new Circle();
      assert.equal(c.radius, 1);
      assert.equal(c.center.toString(), "(0, 0, 0)");
    });
  });

  describe("area", () => {
    it("returns π·r²", () => {
      const c = new Circle(new Vector3(), 5);
      assert.ok(Math.abs(c.area() - Math.PI * 25) < 1e-10);
    });
  });

  describe("perimeter", () => {
    it("returns 2·π·r", () => {
      const c = new Circle(new Vector3(), 3);
      assert.ok(Math.abs(c.perimeter() - 2 * Math.PI * 3) < 1e-10);
    });
  });

  describe("centroid", () => {
    it("equals the center", () => {
      const center = new Vector3(1, 2, 0);
      const c = new Circle(center, 4);
      assert.equal(c.centroid(), center);
    });
  });

  describe("describe", () => {
    it("includes radius and center", () => {
      const c = new Circle(new Vector3(1, 2, 0), 5, Color.RED);
      const d = c.describe();
      assert.ok(d.includes("radius=5"));
      assert.ok(d.includes("(1, 2, 0)"));
      assert.ok(d.includes("RED"));
    });
  });

  describe("getters", () => {
    it("exposes radius and center", () => {
      const center = new Vector3(3, 4, 0);
      const c = new Circle(center, 7, Color.GREEN);
      assert.equal(c.radius, 7);
      assert.equal(c.center, center);
      assert.equal(c.name, "Circle");
    });
  });
});

File: examples/ruby/tests/circle_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "../src/circle"

class CircleTest < Minitest::Test
  def test_area
    c = Circle.new(5)
    assert_in_delta Math::PI * 25, c.area, 1e-9
  end

  def test_perimeter
    c = Circle.new(3)
    assert_in_delta 2 * Math::PI * 3, c.perimeter, 1e-9
  end

  def test_scale
    c = Circle.new(2).scale(3)
    assert_in_delta 6.0, c.radius, 1e-9
  end

  def test_comparable
    small = Circle.new(1)
    large = Circle.new(5)
    assert_operator small, :<, large
  end

  def test_position_default
    c = Circle.new(1)
    assert_equal Vector3::ZERO, c.position
  end

  def test_position_custom
    pos = Vector3.new(1, 2, 0)
    c = Circle.new(1, position: pos)
    assert_equal pos, c.position
  end

  def test_translate
    c = Circle.new(3)
    moved = c.translate(Vector3.new(5, 0, 0))
    assert_equal Vector3.new(5, 0, 0), moved.position
    assert_in_delta c.area, moved.area, 1e-9
  end

  def test_registry
    assert_equal Circle, Shape.registry[:circle]
  end

  def test_to_s
    c = Circle.new(2.5)
    assert_match(/Circle/, c.to_s)
  end
end
